presto-main/src/main/java/com/facebook/presto/operator/BigintInMemoryJoinHash.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.Block;
+import com.google.common.primitives.Ints;
+import io.airlift.slice.XxHash64;
+import it.unimi.dsi.fastutil.HashCommon;
+import it.unimi.dsi.fastutil.longs.LongArrayList;
+
+import java.util.Arrays;
+import java.util.List;
+
+import static com.facebook.presto.operator.SyntheticAddress.decodePosition;
+import static com.facebook.presto.operator.SyntheticAddress.decodeSliceIndex;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.util.Objects.requireNonNull;
+
+// This implementation assumes arrays used in the hash are always a power of 2
+public final class BigintInMemoryJoinHash
+        implements LookupSource
+{
+    private final LongArrayList addresses;
+    private final PagesHashStrategy pagesHashStrategy;
+
+    private final int channelCount;
+    private final int mask;
+    private final int[] key;
+    private final int[] positionLinks;
+    private final long size;
+    private final long[] values;
+
+    public BigintInMemoryJoinHash(LongArrayList addresses, PagesHashStrategy pagesHashStrategy, List<List<Block>> channels, List<Integer> joinChannels)
+    {
+        checkState(requireNonNull(joinChannels, "joinChannels is null").size() == 1, "Only single channel joins are supported");
+        int joinChannel = joinChannels.get(0);
+        this.addresses = requireNonNull(addresses, "addresses is null");
+        this.pagesHashStrategy = requireNonNull(pagesHashStrategy, "pagesHashStrategy is null");
+        this.channelCount = pagesHashStrategy.getChannelCount();
+
+        // reserve memory for the arrays
+        int hashSize = HashCommon.arraySize(addresses.size(), 0.75f);
+
+        mask = hashSize - 1;
+        key = new int[hashSize];
+        Arrays.fill(key, -1);
+
+        positionLinks = new int[addresses.size()];
```

Comment:
postionLinks can be null in case of distinct join values, you can also add this to regular join class